README
My own C library.

Nothing special, just functions I needed...

Requires C11 compiler, e.g. GCC >= 4.7.

-- dennis(a)yurichev.com



ChangeLog
2018-03-05 Dennis Yurichev <dennis(a)yurichev.com>

	* MSVC 2015 support.
	limit in dmalloc (not tested, though).

2017-11-09 Dennis Yurichev <dennis(a)yurichev.com>

	* New util: replace_util.

2017-10-26 Dennis Yurichev <dennis(a)yurichev.com>

	* New functions: L_print_buf_ofs_fds_C(), L_print_buf_ofs_C().
	Dump_util.

2017-10-09 Dennis Yurichev <dennis(a)yurichev.com>

	* New functions: my_strnicmp(), my_truncate_or_die()

2017-10-08 Dennis Yurichev <dennis(a)yurichev.com>

	* MSVC 2015 support

2017-10-07 Dennis Yurichev <dennis(a)yurichev.com>

	* New functions: swap_endianness16(), swap_endianness32()

2017-10-06 Dennis Yurichev <dennis(a)yurichev.com>

	* New function: obj_to_strbuf()
	* No more autotools!

2017-07-25 Dennis Yurichev <dennis(a)yurichev.com>

	* New functions: mulmns(), modinv32(), modinv64().
	* New files: octomath.(h|c)

2017-07-16 Dennis Yurichev <dennis(a)yurichev.com>
	
	* New function: is_prime().

2017-07-15 Dennis Yurichev <dennis(a)yurichev.com>
	
	* New function: uint32_is_2n().

2017-02-22 Dennis Yurichev <dennis(a)yurichev.com>

	* New function: rand_buf().

2017-02-07 Dennis Yurichev <dennis(a)yurichev.com>

	* tetrabyte renamed to tetra and octabyte renamed to octa,
	after names of datatypes in Donald Knuth's MMIX simulator.

2017-01-30 Dennis Yurichev <dennis(a)yurichev.com>
	
	* New function: neat_list_of_bytes().

2016-11-23 Dennis Yurichev <dennis(a)yurichev.com>
	
	* New functions: IsInteger(), uint64_is_2n().

2016-09-28 Dennis Yurichev <dennis(a)yurichev.com>
	
	* New function: entropy_int().

2016-08-28 Dennis Yurichev <dennis(a)yurichev.com>

	* New functions: get_file_size_or_die(), open_or_die(),
	fprint_shrinked_string(), dmalloc_and_snprintf(),
	is_string_consists_only_of_Latin_letters(),
	is_string_consists_only_of_hex_digits(),
	is_string_has_only_one_character_repeating(), CRC64(), popcnt64().

	* New modules: entropy.c, base64.c

2016-08-25 Dennis Yurichev <dennis(a)yurichev.com>

	* New function: rand_bernoulli_distribution()

2016-08-24  Dennis Yurichev <dennis(a)yurichev.com>

	* New functions: ipow(), rotr8(), rotl8(), rotr16(), rotl16().

2016-08-23  Dennis Yurichev <dennis(a)yurichev.com>

	* New functions: string_remove_part(), split_fname()

2016-08-21  Dennis Yurichev <dennis(a)yurichev.com>

	* New function: popcnt32()

2016-04-17  Dennis Yurichev <dennis(a)yurichev.com>
	
	* New functions: regmatch_len(), regmatch_dup(), regexec_to_array_of_string(), dstrndup(), read_text_file_by_line_or_die(),
	streq(), strieq(), print_array_of_strings(), dfree_array_of_blocks(), uint64_log2()

2015-12-08  Dennis Yurichev <dennis(a)yurichev.com>
	
	* New functions: text_file_to_list(), split_list_into_sublists(), delete_if(), nth(), list_pick_random(), print_list_of_strings()

2015-10-04  Dennis Yurichev <dennis(a)yurichev.com>

	* New function: genrand64()

2015-09-06  Dennis Yurichev <dennis(a)yurichev.com>

	* New function: is_buf_printable()

2015-04-09  Dennis Yurichev <dennis(a)yurichev.com>

	* New functions: kmp_search(), XOR_block()

2015-04-06  Dennis Yurichev <dennis(a)yurichev.com>

	* New functions: range_update(), REG_in_range(), REG_in_range2().
	New structure: my_range.

2015-04-05  Dennis Yurichev <dennis(a)yurichev.com>

	* New LISP-like functions: add_to_list(), obj_dup()
	Others: value_not_in2()

2015-02-20  Dennis Yurichev <dennis(a)yurichev.com>

	* Switching to GNU autotools

2013-09-26  Dennis Yurichev <dennis(a)yurichev.com>

	* This is a first public version, uploaded to github.com.


x86.c
/*
 *             _        _   _                           
 *            | |      | | | |                          
 *   ___   ___| |_ ___ | |_| |__   ___  _ __ _ __   ___ 
 *  / _ \ / __| __/ _ \| __| '_ \ / _ \| '__| '_ \ / _ \
 * | (_) | (__| || (_) | |_| | | | (_) | |  | |_) |  __/
 *  \___/ \___|\__\___/ \__|_| |_|\___/|_|  | .__/ \___|
 *                                          | |         
 *                                          |_|
 *
 * Written by Dennis Yurichev <dennis(a)yurichev.com>, 2013
 *
 * This work is licensed under the Creative Commons Attribution-NonCommercial-NoDerivs 3.0 Unported License. 
 * To view a copy of this license, visit http://creativecommons.org/licenses/by-nc-nd/3.0/.
 *
 */

#include <assert.h>

#include "datatypes.h"

#include "x86.h"
#include <ctype.h>
#include "bitfields.h"

#ifdef __GNUC__ 
#include <cpuid.h>
#endif

void flags_to_str (tetrabyte flags, strbuf *out)
{
    u_EFLAGS uf;
    uf.flags=flags;

    if (uf.s.CF) strbuf_addstr (out, "CF ");

    if (uf.s.PF) strbuf_addstr (out, "PF ");
    if (uf.s.AF) strbuf_addstr (out, "AF ");
    if (uf.s.ZF) strbuf_addstr (out, "ZF ");
    if (uf.s.SF) strbuf_addstr (out, "SF ");
    if (uf.s.TF) strbuf_addstr (out, "TP "); // aka TF
    if (uf.s.IF) strbuf_addstr (out, "IF ");
    if (uf.s.DF) strbuf_addstr (out, "DF ");
    if (uf.s.OF) strbuf_addstr (out, "OF ");
    //FIXME:
    //if (IS_SET (flags, 1<<12)) { if (r.size()) r+=" "; r+="IOPL_L"; };
    //if (IS_SET (flags, 1<<13)) { if (r.size()) r+=" "; r+="IOPL_H"; };
    if (uf.s.NT) strbuf_addstr (out, "NT ");
    if (uf.s.RF) strbuf_addstr (out, "RF ");
    if (uf.s.VM) strbuf_addstr (out, "VM ");
    if (uf.s.AC) strbuf_addstr (out, "AC ");
    if (uf.s.VIF) strbuf_addstr (out, "VIF ");
    if (uf.s.VIP) strbuf_addstr (out, "VIP ");
    if (uf.s.ID) strbuf_addstr (out, "ID ");

    if (out->strlen>0)
        strbuf_trim_last_char (out);
};

void dump_flags (fds *s, tetrabyte flags)
{
    strbuf sb = STRBUF_INIT;
    flags_to_str(flags, &sb);

    L_fds (s, sb.buf);

    strbuf_deinit (&sb);
};

void DR7_to_str (tetrabyte DR7, strbuf* out)
{
    strbuf_addf(out, "%s%s%s%s%s%s%s%s%s%s%sR/W0=%d%d LEN0=%d%d R/W1=%d%d LEN1=%d%d R/W2=%d%d LEN2=%d%d R/W3=%d%d LEN3=%d%d", 
            IS_SET (DR7, FLAG_DR7_L0) ? "L0 " : "   ",
            IS_SET (DR7, FLAG_DR7_G0) ? "G0 " : "   ",
            IS_SET (DR7, FLAG_DR7_L1) ? "L1 " : "   ",
            IS_SET (DR7, FLAG_DR7_G1) ? "G1 " : "   ",
            IS_SET (DR7, FLAG_DR7_L2) ? "L2 " : "   ",
            IS_SET (DR7, FLAG_DR7_G2) ? "G2 " : "   ",
            IS_SET (DR7, FLAG_DR7_L3) ? "L3 " : "   ",
            IS_SET (DR7, FLAG_DR7_G3) ? "G3 " : "   ",
            IS_SET (DR7, 1<<8) ? "LE " : "   ",
            IS_SET (DR7, 1<<9) ? "GE " : "   ",
            IS_SET (DR7, 1<<13) ? "GD " : "   ",

            IS_SET (DR7, 1<<16) ? 1 : 0, IS_SET (DR7, 1<<17) ? 1 : 0,  // R/W0
            IS_SET (DR7, 1<<18) ? 1 : 0, IS_SET (DR7, 1<<19) ? 1 : 0,  // LEN0
            IS_SET (DR7, 1<<20) ? 1 : 0, IS_SET (DR7, 1<<21) ? 1 : 0,  // R/W1
            IS_SET (DR7, 1<<22) ? 1 : 0, IS_SET (DR7, 1<<23) ? 1 : 0,  // LEN1
            IS_SET (DR7, 1<<24) ? 1 : 0, IS_SET (DR7, 1<<25) ? 1 : 0,  // R/W2
            IS_SET (DR7, 1<<26) ? 1 : 0, IS_SET (DR7, 1<<27) ? 1 : 0,  // LEN2
            IS_SET (DR7, 1<<28) ? 1 : 0, IS_SET (DR7, 1<<29) ? 1 : 0,  // R/W3
            IS_SET (DR7, 1<<30) ? 1 : 0, IS_SET (DR7, 1<<31) ? 1 : 0); // LEN3
};

void dump_DR7 (fds* s, tetrabyte DR7)
{
    strbuf sb=STRBUF_INIT;
    DR7_to_str(DR7, &sb);
    L_fds (s, sb.buf);
    strbuf_deinit(&sb);
};

void dump_DR6 (fds* s, tetrabyte DR6)
{
    if (IS_SET(DR6, FLAG_DR6_B0)) L_fds (s, "B0 ");
    if (IS_SET(DR6, FLAG_DR6_B1)) L_fds (s, "B1 ");
    if (IS_SET(DR6, FLAG_DR6_B2)) L_fds (s, "B2 ");
    if (IS_SET(DR6, FLAG_DR6_B3)) L_fds (s, "B3 ");
    if (IS_SET(DR6, FLAG_DR6_BS)) L_fds (s, "BS ");
};

void MXCSR_to_str (tetrabyte a, strbuf *out)
{
    if (IS_SET (a, 1<<15)) strbuf_addstr(out, "FZ ");

    if (IS_SET (a, 1<<14)==0 && IS_SET (a, 1<<13)==0) strbuf_addstr(out, "RN ");
    else
        if (IS_SET (a, 1<<14)==1 && IS_SET (a, 1<<13)==1) strbuf_addstr(out, "RZ ");
        else
            if (IS_SET (a, 1<<14)) strbuf_addstr(out, "R+ ");
            else
                if (IS_SET (a, 1<<13)) strbuf_addstr(out, "R- ");

    if (IS_SET (a, 1<<12)) strbuf_addstr(out, "PM ");
    if (IS_SET (a, 1<<11)) strbuf_addstr(out, "UM ");
    if (IS_SET (a, 1<<10)) strbuf_addstr(out, "OM ");
    if (IS_SET (a, 1<<9)) strbuf_addstr(out, "ZM ");
    if (IS_SET (a, 1<<8)) strbuf_addstr(out, "DM ");
    if (IS_SET (a, 1<<7)) strbuf_addstr(out, "IM ");
    if (IS_SET (a, 1<<6)) strbuf_addstr(out, "DAZ ");
    if (IS_SET (a, 1<<5)) strbuf_addstr(out, "PE ");
    if (IS_SET (a, 1<<4)) strbuf_addstr(out, "UE ");
    if (IS_SET (a, 1<<3)) strbuf_addstr(out, "OE ");
    if (IS_SET (a, 1<<2)) strbuf_addstr(out, "ZE ");
    if (IS_SET (a, 1<<1)) strbuf_addstr(out, "DE ");
    if (IS_SET (a, 1<<0)) strbuf_addstr(out, "IE ");

    // TODO trim right part
};

void FCW_to_str (uint16_t a, strbuf *out)
{
    if (IS_SET (a, 1<<12)) strbuf_addstr(out, "IC ");
    switch ((a>>10)&3)
    {
        case 0: strbuf_addstr(out, "RC=NEAR "); break;
        case 1: strbuf_addstr(out, "RC=DOWN "); break;
        case 2: strbuf_addstr(out, "RC=UP "); break;
        case 3: strbuf_addstr(out, "RC=ZERO "); break;
        default: assert (0);
    };

    switch ((a>>8)&3)
    {
        case 0: strbuf_addstr(out, "PC=24bits "); break;
        case 1: strbuf_addstr(out, "PC=?? "); break;
        case 2: strbuf_addstr(out, "PC=53bits "); break;
        case 3: strbuf_addstr(out, "PC=64bits "); break;
        default: assert (0);
    };

    if (IS_SET (a, 1<<7)) strbuf_addstr(out, "IEM ");
    if (IS_SET (a, 1<<5)) strbuf_addstr(out, "PM ");
    if (IS_SET (a, 1<<4)) strbuf_addstr(out, "UM ");
    if (IS_SET (a, 1<<3)) strbuf_addstr(out, "OM ");
    if (IS_SET (a, 1<<2)) strbuf_addstr(out, "ZM ");
    if (IS_SET (a, 1<<1)) strbuf_addstr(out, "DM ");
    if (IS_SET (a, 1<<0)) strbuf_addstr(out, "IM ");

    // TODO trim right part
};

void FSW_to_str (uint16_t a, strbuf *out)
{
    if (IS_SET (a, 1<<15)) strbuf_addstr(out, "B ");
    if (IS_SET (a, 1<<14)) strbuf_addstr(out, "C3 ");
    strbuf_addf (out, "TOP=%d ", FPU_TOP(a));
    if (IS_SET (a, 1<<10)) strbuf_addstr(out, "C2 ");
    if (IS_SET (a, 1<<9)) strbuf_addstr(out, "C1 ");
    if (IS_SET (a, 1<<8)) strbuf_addstr(out, "C0 ");
    if (IS_SET (a, 1<<7)) strbuf_addstr(out, "IR ");
    if (IS_SET (a, 1<<6)) strbuf_addstr(out, "SF ");
    if (IS_SET (a, 1<<5)) strbuf_addstr(out, "P ");
    if (IS_SET (a, 1<<4)) strbuf_addstr(out, "U ");
    if (IS_SET (a, 1<<3)) strbuf_addstr(out, "O ");
    if (IS_SET (a, 1<<2)) strbuf_addstr(out, "Z ");
    if (IS_SET (a, 1<<1)) strbuf_addstr(out, "D ");
    if (IS_SET (a, 1<<0)) strbuf_addstr(out, "I ");

    // TODO trim right part
};

void XMM_to_strbuf (byte* p, strbuf *sb)
{
    int i;
    byte a;

    strbuf_addstr(sb, "0x");
    //	double d;
    //	float f;

    for (i=0; i<16; i++)
        strbuf_addf (sb, "%02X", *(p+(15-i)));
    // ASCII
    strbuf_addstr(sb, "   \"");
    for (i=0; i<16; i++)
    {
        a=*(p+i); // string is in reverse order
        if (isprint (a))
            strbuf_addf (sb, "%c", a);
        else
            strbuf_addc (sb, '.');
    };

    strbuf_addc (sb, '\"');

    //	memcpy (&d, p, sizeof (d)); // low part of XMM register
    //	memcpy (&f, p, sizeof (f)); // low part of XMM register

    //	if (_isnan(d)==0)
    //		rt+=strfmt ("lowpart=%lf", d);
};

bool sse_supported()
{
#ifdef _MSC_VER
    int b[4];
    __cpuid(b,1);
    if (b[3] & (1<<25)) // EDX, bit 25
        return true;
#else
    int a, b, c, d;
    __cpuid(1, a, b, c, d);
    if (c & (1<<25)) // EDX, bit 25
        return true;
#endif
    return false;
};

bool sse2_supported()
{
#ifdef _MSC_VER
    int b[4];
    __cpuid(b,1);
    if (b[3] & (1<<26)) // EDX, bit 26
        return true;
#else
    int a, b, c, d;
    __cpuid(1, a, b, c, d);
    if (c & (1<<26)) // EDX, bit 26
        return true;
#endif
    return false;
};

/* vim: set expandtab ts=4 sw=4 : */
